#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/dji/mavic/protos/Mavic2Pro.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/SandyGround.proto"

WorldInfo {
  basicTimeStep 16
  defaultDamping Damping {
    linear 0.5
    angular 0.3
  }
}
Viewpoint {
  orientation -0.16441195249034138 0.7605679283593586 0.6280964386377839 0.6629548685257637
  position -37.36795567035433 -20.09272460024777 27.968649084273117
  followType "None"
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Floor {
  size 50 50
  appearance SandyGround {
  }
}
Robot {
  translation 0 0 0.0399049
  children [
    Receiver {
      name "receiver_ground"
    }
    Emitter {
      name "emitter_ground"
    }
    GPS {
      translation 0 0 0.005
      children [
        Solid {
          translation 0 0 0.01
          children [
            Shape {
              appearance PBRAppearance {
              }
              geometry Cylinder {
                height 0.02
                radius 0.002
              }
            }
          ]
        }
      ]
      name "gps_ground"
    }
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0.917647 0.917647 0.917647
        baseColorMap ImageTexture {
          url [
            "textures/4x4_1000-0 (Custom).png"
          ]
        }
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.5 0.5 0.05
      }
    }
    DEF WHEEL1 HingeJoint {
      jointParameters HingeJointParameters {
        position 9.347513871526907e-09
        axis 0 1 0
        anchor 0.18 0.26 0
      }
      device [
        RotationalMotor {
          name "wheel1"
        }
      ]
      endPoint Solid {
        translation 0.18 0.26 0
        rotation 1 4.871026233501893e-09 -4.871047169352938e-09 1.5707999999999502
        children [
          DEF WHEEL Shape {
            appearance PBRAppearance {
              baseColor 0.305882 0.898039 0.25098
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.02
              radius 0.04
              subdivision 24
            }
          }
        ]
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL2 HingeJoint {
      jointParameters HingeJointParameters {
        position 1.0049746295156312e-08
        axis 0 1 0
        anchor 0.18 -0.26 0
      }
      device [
        RotationalMotor {
          name "wheel2"
        }
      ]
      endPoint Solid {
        translation 0.18 -0.26 0
        rotation 1 5.0248542977002495e-09 -5.024872755055763e-09 1.570799999999945
        children [
          USE WHEEL
        ]
        name "solid(1)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL3 HingeJoint {
      jointParameters HingeJointParameters {
        position -1.183077672291902e-08
        axis 0 1 0
        anchor -0.18 0.26 0
      }
      device [
        RotationalMotor {
          name "wheel3"
        }
      ]
      endPoint Solid {
        translation -0.18 0.26 0
        rotation 1 -6.16166485760548e-09 6.161687292540311e-09 1.5707999999999513
        children [
          USE WHEEL
        ]
        name "solid(2)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL4 HingeJoint {
      jointParameters HingeJointParameters {
        position -1.1689693413617427e-08
        axis 0 1 0
        anchor -0.18 -0.26 0
      }
      device [
        RotationalMotor {
          name "wheel4"
        }
      ]
      endPoint Solid {
        translation -0.18 -0.26 0
        rotation 1 -5.84482405996045e-09 5.844847491853501e-09 1.570799999999959
        children [
          USE WHEEL
        ]
        name "solid(3)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF DS_RIGHT DistanceSensor {
      translation 0.16 -0.03 0
      rotation 0 0 1 -0.3
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.184314 0.596078 0.847059
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_right"
    }
    DEF DS_LEFT DistanceSensor {
      translation 0.16 0.03 0
      rotation 0 0 1 0.3
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.184314 0.596078 0.847059
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_left"
    }
  ]
  boundingObject USE BODY
  physics Physics {
    density -1
    mass 1
  }
  controller "ground_controller"
  supervisor TRUE
}
Mavic2Pro {
  translation 0 0 0.135
  controller "mavic_controller"
  cameraSlot [
    Camera {
      width 800
      height 600
      exposure 5
      antiAliasing TRUE
    }
  ]
}
